empiric-package/empiric/core/mixins/nonce.py

```diff
@@ -0,0 +1,101 @@
+from typing import Dict, Literal, Optional, Union
+
+from starknet_py.net.client import Client
+from starknet_py.net.client_models import TransactionReceipt, TransactionStatus
+
+
+class NonceMixin:
+    client: Client
+    account_contract_address: Optional[int]
+    nonce_status: Dict[int, TransactionStatus] = dict()
+    nonce_dict: Dict[int, int] = dict()
+    pending_nonce: Optional[int] = None
+
+    async def _get_nonce(self) -> int:
+        self.update_nonce_dict()
+        self.cleanup_nonce_dict()
+
+        if self.pending_nonce:
+            return self.pending_nonce
+        if self.nonce_status:
+            return max(self.nonce_dict) + 1
+
+        return await self.get_latest_nonce()
+
+    def cleanup_nonce_dict(self):
+        nonce_seq = [
+            x
+            for x in self.nonce_dict
+            if self.nonce_status[x]
+            in [TransactionStatus.ACCEPTED_ON_L1, TransactionStatus.ACCEPTED_ON_L2]
+        ]
+        if nonce_seq:
+            max_accepted = max(nonce_seq)
+            for nonce in list(self.nonce_dict):
+                if nonce <= max_accepted:
+                    del self.nonce_dict[nonce]
+                    if nonce in self.nonce_status:
+                        del self.nonce_status[nonce]
+            if self.pending_nonce and self.pending_nonce < max_accepted:
+                self.pending_nonce = None
+
+    async def track_nonce(
+        self,
+        transaction,
+        response,
+    ):
+        nonce = transaction.nonce
+        transaction_hash = response.transaction_hash
+
+        self.nonce_dict[nonce] = transaction_hash
+        self.nonce_status[nonce] = await self.get_status(transaction_hash)
+
+        nonce_min = min(self.nonce_dict)
+        nonce_max = max(self.nonce_dict)
+        for i in range(nonce_min, nonce_max + 1):
+            if i not in self.nonce_dict:
+                self.pending_nonce = i
+                break
+
+    async def update_nonce_dict(
+        self,
+    ):
+        for nonce in list(self.nonce_dict):
+            self.nonce_status[nonce] = await self.get_status(self.nonce_dict[nonce])
+            if self.nonce_dict[nonce] in [
+                # TODO: we should prob confirm received when sending
+                # TransactionStatus.NOT_RECEIVED,
+                TransactionStatus.REJECTED,
+            ]:
+                self.pending_nonce = min(self.pending_nonce, nonce)
+                del self.nonce_dict[nonce]
```

Comment:
I think here we should just reset the `nonce_dict` because all previous nonces must have been accepted and all later must be invalid.

Reply:
More generally I think we can simplify the logic: Keep an ordered list of past nonces that were sent in `nonce_dict`: nonce -> tx_hash mapping. When sending a new tx start at the top of the list and check the status for the first nonce. If it is received, exit. If it is accepted on L1 or L2, remove it from the list and proceed to the next element in the list. If it is rejected, reset the nonce to the nonce of that tx and clear the `nonce_dict`.

Reply:
i clean up the nonce_dict and nonce_status if there's an error now